### DISCUSSION TITLE
Nullable pattern for custom type

### DISCUSSION THREAD

**Discussion by Thaina** [dotnet/csharplang#494] [General]
This is an attempt to generalized #147 

Which I was seen it has benefit for an argument about `default` in #99 

## Motivation

Sometimes we create custom type that can be in state that should not be used. One good example is `Fraction` struct

```C#
struct Fraction
{
    public long Num,Div;
    public double Value { get => Num / (double)Div }
}
```

Initially this struct would made into 0/0. Which is NaN and not an initial value it should be used

So it more like this struct is inherently nullable struct by its own state. If we could treat it as `null value` then there would be helper operators (exist and will be exist) that could be reused. Especially `??` and `?.`

## Nullable Pattern

I suggest that we should have `INullable` interface to made a contract with all null related operator

But it might be the same as `IEnumerable` that it would not really need to define type with this interface. Just implement the method with same name (but to have interface would prevent breaking change)

```C#
interface INullable { bool HasValue { get; } }
struct Fraction : INullable
{
    public long Num,Div;
    public double Value { get => Num / (double)Div }
    public bool HasValue { get => !double.IsNan(Value); }
}

var fraction = new Fraction(); // Initial 0/0 is NaN
var isNull = fraction == null; // true
var value = fraction?.Value ?? 1.0; // 1.0
```

I was using `Fraction` as an example but it could be class too. For example, disposable object such as unity's `GameObject` would also use this pattern for destroyed gameobject

## Alternative

It would be easier if we could make it with static operator syntax

```C#
struct Fraction
{
    public long Num,Div;
    public double Value { get => Num / (double)Div }

    // default null value
    public static implicit Fraction operator null() => default(Fraction);
    public static implicit bool operator null(ref Fraction f) => !double.IsNan(f.Value);
    // with this it could became null with any of field is NaN too
}

/// Work the same as above
/// operator name "null" return bool if it null
```

## Related issue

#146,#147 Origin of this issue
also https://github.com/dotnet/roslyn/issues/4347

#99 Motivation to solve problem about `default` struct and constructed struct
This feature does not make us avoid uninitialized struct like #99 have argue. But it generalized a way we could explicitly know that an object of any types is in `undefined` state

#34 `??=` Would work well on this

#36 `!` operator could also work on this. Then it would need #99 and also solve #99 problem of initialize array of struct

```C#
var nullableFractions = new Fraction[10]; // fast create zeromemory block
var nonNullFractions = new Fraction![10]; // calling default constructor on each block
```

By explicitly use `!` it then preserve the original behaviour

## drawback

This feature would be hard to let API consumer know that which struct could be nullable and which can't

## other

I was mention a `default null value` above. Because I think we should be able to assign null value to this object if it is struct. And it would be that returned value

**Comment by lachbaer** [comment]
I downvoted on the first idea and the drawbacks in general. But I upvoted on the basic - though not concrete - idea of the `operator null`, because it has similarities to a solution I'm currently validating (for mine I have `operator ?` in mind, because it will work with all current `?` constructs and simultaneously be an unary operator of its own. But please wait for the someday upcoming proposal).
